atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java

```diff
@@ -1681,125 +1627,126 @@ private static boolean isSweepSentinel(Value value) {
+                .removeKeys(keysToReload::containsKey)
+                .removeValues(value -> value.getContents().length == 0)
+                // The value has a commit timestamp less than our start timestamp, and is visible and valid.
+                .collect(MoreCollectors.entriesToCustomMap(LinkedHashMap::new));
+
+        EntryStream<Cell, Value> resultStream = EntryStream.of(filteredResults);
+        if (shouldDeleteAndRollback()) {
+            // These are from a failed transaction so we can roll it back and then reload it.
+            getCounter(AtlasDbMetricNames.CellFilterMetrics.INVALID_COMMIT_TS, tableRef)
+                    .inc(keysToDelete.size());
+            if (!keysToDelete.isEmpty()
+                    && !rollbackFailedTransactions(
+                            tableRef, keysToDelete, commitTimestamps, defaultTransactionService)) {
+                return resultStream.append(getWithPostFilteringInternal(
+                        tableRef,
+                        Maps.difference(rawResults, filteredResults).entriesOnlyOnLeft(),
+                        asyncKeyValueService,
+                        asyncTransactionService,
+                        iterations + 1));
             }
         }
 
-        if (!keysToReload.isEmpty()) {
-            return Futures.transform(
-                    asyncKeyValueService.getAsync(tableRef, keysToReload),
-                    nextRawResults -> {
-                        boolean allPossibleCellsReadAndPresent = nextRawResults.size() == keysToReload.size();
+        Map<Cell, Long> keysToReloadOrDelete = ImmutableMap.<Cell, Long>builderWithExpectedSize(
+                        keysToReload.size() + keysToDelete.size())
+                .putAll(keysToReload)
+                .putAll(keysToDelete)
+                .buildOrThrow();
+        if (!keysToReloadOrDelete.isEmpty()) {
+            return resultStream.append(streamFutureEntries(asyncKeyValueService
+                    .getCompletableAsync(tableRef, keysToReloadOrDelete)
+                    .thenApply(nextRawResults -> {
                         validatePreCommitRequirementsOnReadIfNecessary(
-                                tableRef, getStartTimestamp(), allPossibleCellsReadAndPresent);
-                        return getRemainingResults(nextRawResults, keysAddedToResults);
-                    },
-                    MoreExecutors.directExecutor());
+                                tableRef, getStartTimestamp(), nextRawResults.size() == keysToReloadOrDelete.size());
+                        return Maps.difference(rawResults, filteredResults).entriesOnlyOnLeft();
+                    })
+                    .thenApply(remainingRawResults -> getWithPostFilteringInternal(
+                            tableRef,
+                            remainingRawResults,
+                            asyncKeyValueService,
+                            asyncTransactionService,
+                            iterations + 1))
```

Comment:
recursion ⚠️

Reply:
there's a max iterations precondition already so no danger of stack overflow